(TK-173) Support querying based on status-version

MSLilah opened the pull request:
Add support for querying based on a specific version of a
service's registered status function using the
service-status-version parameter.

MSLilah committed:
(TK-173) Support querying based on status-version

Add support for querying based on a specific version of a
service's registered status function using the
service-status-version parameter.

cprice404 commented on src/puppetlabs/trapperkeeper/services/status/ringutils.clj:
```diff
@@ -14,7 +14,8 @@
     (try+ (handler request)
-          (catch [:type :request-data-invalid] e
+          (catch #(or (= :request-data-invalid (:type %))
+                      (= :service-status-version-not-found (:type %))) e
```
This could be `#(contains? #{:request-data-invalid :service-status-version-not-found} (:type %))`, but, that's not a blocker

cprice404 commented on src/puppetlabs/trapperkeeper/services/status/status_core.clj:
```diff
@@ -45,23 +45,40 @@
+      (when (nil? status)
+        (throw+ {:type    :service-status-version-not-found
+                 :message (str "No status function with version "
+                               service-status-version
```
this is going to log a `nil` if they didn't specify a status version and there are no callbacks registered for the service, correct?  Or do we already have a guard elsewhere that guarantees that we won't get here if they haven't registered a callback?

rlinehan replied:
It will only be `nil` if a service version wasn't supplied. In that case, there must be at least one status function registered for the service, otherwise we'll return a `404` because we can't find anything for the service name. So I think this is fine.

cprice404 replied:
k

cprice404 commented on src/puppetlabs/trapperkeeper/services/status/status_core.clj:
```diff
@@ -76,6 +93,18 @@
+  (when-let [level (params :service-status-version)]
+    (try
+      (Integer. level)
+      (catch NumberFormatException e
```
I think we have a parse-int fn in kitchensink that does basically this, but if we don't already have a dep on KS it's probably not worth introducing one just for that.

rlinehan replied:
Given that trapperkeeper itself has a dependency on KS, won't we always have a KS dependency? Or are you saying we don't want to have a dependency on a specific version of KS if we don't have to?

cprice404 replied:
oh, duh.  Yeah probably worth just using the fn from ks then.

cprice404 commented:
LGTM, +1

rlinehan commented:
:+1: looks good! Seems like it would be slightly easier to make the two small changes that @cprice404 mentioned before merging, but we could also open up a maint PR after this gets merged - I'm fine with either.

MSLilah committed:
(TK-173) Address PR comments

Use parse-int function from kitchensink. Cleanup catch clause in
`wrap-request-data-errors`.

MSLilah commented:
@cprice404 @rlinehan I've addressed your comments!

rlinehan commented on src/puppetlabs/trapperkeeper/services/status/status_core.clj:
```diff
@@ -76,6 +93,18 @@
+  [params]
+  (when-let [level (params :service-status-version)]
+    (let [parsed-level (ks/parse-int level)]
+      (if (nil? parsed-level)
```
I think this could be simplified by doing `(if-let [parsed-level (ks/parse-int level)]`

MSLilah commented:
@rlinehan I simplified those lines into an `if-let` statement

rlinehan commented:
dear travis, you are taking way too long.
